src/server/game/Instances/InstanceScript.cpp

```diff
@@ -535,99 +535,87 @@ void InstanceScript::DoRespawnGameObject(ObjectGuid guid, uint32 timeToDespawn /
         TC_LOG_DEBUG("scripts", "InstanceScript: DoRespawnGameObject failed");
 }
 
-void InstanceScript::DoUpdateWorldState(uint32 uiStateId, uint32 uiStateData)
+void InstanceScript::DoOnPlayers(std::function<void(Player*)>&& function)
```

Comment:
My gut disagrees with needing to pun to std::function here.

If this is moved to `Map`, it can be fully in the header without issue, since `Map::PlayerList` is visible in that header.

```c++
template <typename T>
void DoOnPlayers(T&& fn)
{
    for (auto const& ref : GetPlayers())
        if (Player* player = ref.GetSource())
            fn(player);
}
```

PS is it even possible for source to be nullptr there?

Reply:
No, but this mistake is repeated everywhere

Reply:
Sounds like a perfect opportunity to fix this then.

Reply:
Well, there is one case when this can be nullptr and thats when during such a loop, player object is freed (which is never)

Reply:
Aren't player objects only freed at the end of the tick?

Is it possible to cause iterator invalidation during this loop (by teleporting players etc.?)

Reply:
No, none of the things you mentioned happen in: 1) map update 2) a loop iterating player list on map

Reply:
Well, assuming some script does something silly where it does `DoOnPlayers([](Player* p) { p->TeleportTo(wherever); })`